### ISSUE TITLE
Yarn Plug'n'Play support for babel-node

### ISSUE THREAD

**Issue by mischnic** [babel/babel#8967]
## Bug Report / Feature request

**Current Behavior**
Running babel-node with this package.json installed with yarn gives runtime errors.

```
/Users/niklas/Library/Caches/Yarn/v3/npm-@babel-node-7.0.0-20e55bb0e015700a0f6ff281c712de7619ad56f4/node_modules/@babel/node/lib/babel-node.js:99
    if (err.code !== "MODULE_NOT_FOUND") throw err;
                                         ^

Error: Package "@babel/node@7.0.0" (via "/Users/niklas/Library/Caches/Yarn/v3/npm-@babel-node-7.0.0-20e55bb0e015700a0f6ff281c712de7619ad56f4/node_modules/@babel/node/lib/babel-node.js") is trying to require the package "kexec" (via "kexec") without it being listed in its dependencies (@babel/core, @babel/polyfill, @babel/register, commander, fs-readdir-recursive, lodash, output-file-sync, v8flags, @babel/node)
    at makeError (.../.pnp.js:54:17)
    at Object.resolveToUnqualified (.../.pnp.js:3860:17)
    at Object.resolveRequest (.../.pnp.js:3931:31)
    at Function.Module._resolveFilename (.../.pnp.js:4111:30)
    at Function.Module._load (.../.pnp.js:4029:31)
    at Module.require (internal/modules/cjs/loader.js:643:17)
    at require (internal/modules/cjs/helpers.js:22:18)
    at /Users/niklas/Library/Caches/Yarn/v3/npm-@babel-node-7.0.0-20e55bb0e015700a0f6ff281c712de7619ad56f4/node_modules/@babel/node/lib/babel-node.js:94:19
    at /Users/niklas/Library/Caches/Yarn/v3/npm-v8flags-3.1.1-42259a1461c08397e37fe1d4f1cfb59cad85a053/node_modules/v8flags/index.js:131:14
    at /Users/niklas/Library/Caches/Yarn/v3/npm-v8flags-3.1.1-42259a1461c08397e37fe1d4f1cfb59cad85a053/node_modules/v8flags/index.js:39:14
```

Fixing this line to also ignore the error code `UNDECLARED_DEPENDENCY`

https://github.com/babel/babel/blob/d35563ee1a7c3d8a69b5bc1b87fc8d6d883f00ea/packages/babel-node/src/babel-node.js#L91

leads to this error:

```
internal/modules/cjs/loader.js:589
    throw err;
    ^

Error: Cannot find module '@babel/polyfill'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:587:15)
    at Function.Module._load (internal/modules/cjs/loader.js:513:25)
    at Module.require (internal/modules/cjs/loader.js:643:17)
    at require (internal/modules/cjs/helpers.js:22:18)
    at Object.<anonymous> (/Users/niklas/Library/Caches/Yarn/v3/npm-@babel-node-7.0.0-20e55bb0e015700a0f6ff281c712de7619ad56f4/node_modules/@babel/node/lib/_babel-node.js:73:1)
    at Module._compile (internal/modules/cjs/loader.js:707:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:718:10)
    at Module.load (internal/modules/cjs/loader.js:605:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:544:12)
    at Function.Module._load (internal/modules/cjs/loader.js:536:3)
error Command failed with exit code 1.
```

**Expected behavior/code**
No errors

**Babel Configuration (.babelrc, package.json, cli command)**

```json
{
  "private": true,
  "devDependencies": {
    "@babel/core": "^7.0.0",
    "@babel/node": "^7.0.0",
    "@babel/preset-env": "^7.0.0",
    "@babel/preset-react": "^7.0.0"
  },
  "installConfig": {
    "pnp": true
  }
}
```
```json
{
	"presets": [["env", {"target": "current"}], "react"]
}
```

**Environment**
- Babel version(s): @babel/core: 7.1.2
- Node version: v11.0.0
- Yarn: 1.12.1
- OS: macOS 10.13.6
- Monorepo: no
- How you are using Babel: babel-node

@arcanis might be interested

**Comment by xtuc** [comment]
I haven't got the time to look into Yarn Plug'n'Play yet, what should `babel/node` do to make it work? Why is that dependency missing?

Also, could you please share your package.json?

**Comment by arcanis** [comment]
Hey! The `Module not found` error at the end is typically thrown when the hook is not properly setup, which most often occurs when you spawn a node subprocess without forwarding `process.execArgv`.

I've looked a bit in `@babel/node` and there's one place where it happens - fixing it should get you past this message 😃

**Comment by mischnic** [comment]
Changing 

https://github.com/babel/babel/blob/d35563ee1a7c3d8a69b5bc1b87fc8d6d883f00ea/packages/babel-node/src/babel-node.js#L91-L96

to

```js
    if (err.code !== "MODULE_NOT_FOUND" && err.code !== "UNDECLARED_DEPENDENCY") throw err;

    const child_process = require("child_process");

    const proc = child_process.fork(args[0], args.slice(1), {
      stdio: "inherit"
    });
```

gives this new error:

```
~/Library/Caches/Yarn/v3/npm-@babel-core-7.1.2-f8d2a9ceb6832887329a7b60f9d035791400ba4e/node_modules/@babel/core/lib/config/files/plugins.js:152
    throw e;
    ^

Error: Cannot find module '@babel/preset-env' from '.../example-proton-pnp-babel'
    at Function.module.exports [as sync] (~/Library/Caches/Yarn/v3/npm-resolve-1.8.1-82f1ec19a423ac1fbd080b0bab06ba36e84a7a26/node_modules/resolve/lib/sync.js:43:15)
    at resolveStandardizedName (~/Library/Caches/Yarn/v3/npm-@babel-core-7.1.2-f8d2a9ceb6832887329a7b60f9d035791400ba4e/node_modules/@babel/core/lib/config/files/plugins.js:101:31)
    at resolvePreset (~/Library/Caches/Yarn/v3/npm-@babel-core-7.1.2-f8d2a9ceb6832887329a7b60f9d035791400ba4e/node_modules/@babel/core/lib/config/files/plugins.js:58:10)
    at loadPreset (~/Library/Caches/Yarn/v3/npm-@babel-core-7.1.2-f8d2a9ceb6832887329a7b60f9d035791400ba4e/node_modules/@babel/core/lib/config/files/plugins.js:77:20)
    at createDescriptor (~/Library/Caches/Yarn/v3/npm-@babel-core-7.1.2-f8d2a9ceb6832887329a7b60f9d035791400ba4e/node_modules/@babel/core/lib/config/config-descriptors.js:154:9)
    at items.map (~/Library/Caches/Yarn/v3/npm-@babel-core-7.1.2-f8d2a9ceb6832887329a7b60f9d035791400ba4e/node_modules/@babel/core/lib/config/config-descriptors.js:109:50)
    at Array.map (<anonymous>)
    at createDescriptors (~/Library/Caches/Yarn/v3/npm-@babel-core-7.1.2-f8d2a9ceb6832887329a7b60f9d035791400ba4e/node_modules/@babel/core/lib/config/config-descriptors.js:109:29)
    at createPresetDescriptors (~/Library/Caches/Yarn/v3/npm-@babel-core-7.1.2-f8d2a9ceb6832887329a7b60f9d035791400ba4e/node_modules/@babel/core/lib/config/config-descriptors.js:101:10)
    at presets (~/Library/Caches/Yarn/v3/npm-@babel-core-7.1.2-f8d2a9ceb6832887329a7b60f9d035791400ba4e/node_modules/@babel/core/lib/config/config-descriptors.js:47:19)
```

So a `resolve.sync` call here: https://github.com/babel/babel/blob/d35563ee1a7c3d8a69b5bc1b87fc8d6d883f00ea/packages/babel-core/src/config/files/plugins.js#L99

**Comment by mischnic** [comment]
@arcanis The last error comes from the [resolve](https://www.npmjs.com/package/resolve) npm package. But .pnp.js

```diff
```

**Comment by mischnic** [comment]
@arcanis If I change the .pnp.js to also patch the resolve npm package for babel/code, it works (and the fork patch for babel above)!

Is there a reason to not always patch resolve?

```diff
patchedModules.set('resolve', realResolve => {
 const mustBeShimmed = caller => {
   const callerLocator = exports.findPackageLocator(caller);
 
-  return callerLocator && callerLocator.name === 'liftoff';
+  return callerLocator && (callerLocator.name === 'liftoff' || "@babel/core");
 };

```

**Comment by mischnic** [comment]
Why does babel-node try to require `kexec` ? When does that even succeed if it's not listed as a dependency?

**Comment by arcanis** [comment]
Hey! As you might have seen, I've reworked a bit the way the PnP hook is injected, and as a result forwarding the `execArgv` array isn't required anymore (still a good practice, maybe?). This only leaves the problem of `kexec`.

> When does that even succeed if it's not listed as a dependency?

It succeeds accidentally when the hoisting puts kexec to the top-level, or when the user of `babel-node` depends on kexec. In essence, I think the idea is to make it an optional dependency (which is a use case I'm working to solve here: https://github.com/yarnpkg/rfcs/pull/105).

Additionally, it doesn't break when the package isn't found because the require is done in a `try/catch`, *but* it also check for `err.code` being equal to `MODULE_NOT_FOUND`, which isn't the case on PnP (we throw semantic errors, so depending on the reason why the module isn't found it can be a different code such as `UNDECLARED_DEPENDENCY`, `MISSING_PEER_DEPENDENCY`, or `MISSING_DEPENDENCY`).

> Is there a reason to not always shim resolve? (I guess you want to fix resolve itself)

Yup, that's the idea! I have a PR setup (https://github.com/browserify/resolve/pull/170) to expose a better handling of this semantic case. It would still require `babel-node` to use the `useProcessResolution` flag, but that seems acceptable.

Now you mention it, maybe it should be configurable through the yarnrc files 🤔 

That said, there's an easy workaround: just replace your `.babelrc` by `.babelrc.js` and use `require.resolve`. For example:

```js
module.exports = {
  presets: [
    require.resolve(`@babel/preset-env`)
  ]
};
```

This is totally portable 🙂

**Comment by mischnic** [comment]
> forwarding the execArgv array isn't required anymore (still a good practice, maybe?).

Great!

>> When does that even succeed if it's not listed as a dependency?
>
> It succeeds accidentally when the hoisting puts kexec to the top-level, or when the user of babel-node depends on kexec.

Let me rephrase my question: What's the intention here? Was kexec once a dependency and it was accidentally removed? Seems strange that it needs to be depend on by the user (and without hoisting that wouldn't even work)

> In essence, I think the idea is to make kexec an optional dependency

But I think kexec would need be updated because it doesn't compile on Windows and is missing a `os` key in the package.json.

> It would still require babel-node to use the useProcessResolution flag, but that seems acceptable.

Yes, in babel-core (but there are many calls to resolve):
https://github.com/babel/babel/blob/master/packages/babel-core/src/config/files/plugins.js#L99

**Comment by merceyz** [comment]
This can be closed as I can't reproduce with the latest versions of Babel and Yarn in the provided repro
